src/eval/stack.rs

```diff
@@ -132,6 +132,20 @@ impl Stack {
         self.0.clear();
     }
 
+    /// Pops all items in the stack and resets the state of the thunks it encounters.
+    pub fn reset_thunks(&mut self) {
```

Comment:
I like recursion as much as the next functional programmer :stuck_out_tongue: but this will use stack space (the binary process stack, not our heap-allocated `Stack`) linear in the size of `stack` (our `Stack`, `self`).

More idiomatic Rust would be:

```rust
    pub fn reset_thunks(self) {
        for elt in self.0 {
            if let Some(Marker::Thunk(thunk)) = elt {
                thunk.reset_state();
             }
        }
    }
```

Reply:
Your solution is way better. Actually I wanted to write something like this but somehow I was convinced this would cause some problems with moving out of self.0, but self doesn't need to be mutable so there's no problem. My bad for not realizing that (and not trying it out before writing a non-idiomatic solution)

Reply:
Yeah but doesn't this move out stuff behind a mutable reference?

Reply:
Yes, you need to adapt for the borrowing you want.